pkg/controller/operators/olm/operator.go

```diff
@@ -521,19 +624,55 @@ func (a *Operator) crdOwnerConflicts(in *v1alpha1.ClusterServiceVersion, csvsInN
 	return nil
 }
 
-// annotateNamespace is the method that gets called when we see a namespace event in the cluster
-func (a *Operator) annotateNamespace(obj interface{}) (syncError error) {
+// syncNamespace is the method that gets called when we see a namespace event in the cluster
+func (a *Operator) syncNamespace(obj interface{}) (syncError error) {
 	namespace, ok := obj.(*corev1.Namespace)
+	namespaceName := namespace.GetName()
 	if !ok {
 		log.Debugf("wrong type: %#v", obj)
 		return fmt.Errorf("casting Namespace failed")
 	}
 
-	log.Infof("syncing Namespace: %s", namespace.GetName())
+	log.Infof("syncing Namespace: %s", namespaceName)
 	if err := a.annotator.AnnotateNamespace(namespace); err != nil {
-		log.Infof("error annotating namespace '%s'", namespace.GetName())
+		log.Infof("error annotating namespace '%s'", namespaceName)
+		return err
+	}
+
+	opGroupList, err := a.client.OperatorsV1alpha2().OperatorGroups(namespaceName).List(metav1.ListOptions{})
+	if err != nil {
 		return err
 	}
+
+	for _, op := range opGroupList.Items {
+		selector, err := metav1.LabelSelectorAsSelector(&op.Spec.Selector)
+		if err != nil {
+			return err
+		}
+
+		needsUpdate := true
+		if selector.Matches(labels.Set(namespace.GetLabels())) {
+			for _, seenNamespace := range op.Status.Namespaces {
+				if seenNamespace.Name == namespaceName {
+					needsUpdate = false
+					break
+				}
+			}
+		} else {
+			for _, seenNamespace := range op.Status.Namespaces {
+				if seenNamespace.Name == namespaceName {
+					goto DoUpdate
+				}
+			}
+			needsUpdate = false
+		}
+	DoUpdate:
+		if needsUpdate {
+			if err := a.updateDeploymentAnnotation(&op); err != nil {
```

Comment:
this is going to update for every single namespace that's new, I don't think we want it to be in the loop. that way you can just break the loop and don't need a label/goto